### PR TITLE
Fix residuals within pressure solver

### PR DESCRIPTION
**Pull request by weymouth**
This PR gets rid of the undocumented and slightly magical σᵥ divergence correction and replaces it with a documented `residual!` function that corrects the residual at the start of every pressure `solver!` call. This gets rid of the σᵥ field (eliminating 
 its calculation loop and reducing memory a bit) and should improve solver robustness at the expense of putting an `ifelse` statement @inside the residual calculation.

### CODE REVIEW

**Comment by weymouth** [comment]
I ran the benchmarks and the PR is slightly faster, but it's in the noise for me. Can someone else check.

File: src/Poisson.jl [hunk 1]
```diff
@@ -71,7 +71,26 @@ end
...
+`sum(r) = 0` which ensures global satisfiability.
+
+The global correction is done by adjusting all points uniformly, 
+minimizing the local effect. Other approaches are possible.
+
+Note: These corrections mean `x` is not strictly solving `Ax=z`, but
+without the corrections, no solution exists.
+"""
+function residual!(p::Poisson) 
+    @inside p.r[I] = ifelse(p.iD[I]==0,0,p.z[I]-mult(I,p.L,p.D,p.x))
```
Comment (b-fg) [inline src/Poisson.jl:89]: Shouldn't the `ifelse` be converted for floating point types? Something like:
```
    @inside p.r[I] = ifelse(p.iD[I]≈zeroT,zeroT,p.z[I]-mult(I,p.L,p.D,p.x))
```
And then we can define `zeroT` in `util.jl` according to the `T` type, and export it. I could push a PR for this separately if you'd like. Would be nice to also clean this up in other places.

**Comment by b-fg** [comment]
I think the benchmark should be tested on a flexible/volume-changing geometry. Should I include the jellyfish simulation in the benchmark folder and test this?

File: src/Poisson.jl (hunk 1 above)
Comment (weymouth) [reply src/Poisson.jl:89]: Sure. But why not just using `zero(T)`? I doubt the performance impact will be enough for me to measure on my (noisy) laptop.

**Comment by weymouth** [comment]
I agree that it's probably good to get a _dynamic_ geometry case in the benchmarks. I think the jelly is the only _size changing_ body case, but that may not be critical. You could use the shark case or the hover case if you prefer.

**Comment by b-fg** [comment]
Those are 2D, which would need to be quite big then to fill the GPU. Hopefully that would not be a problem.

**Comment by weymouth** [comment]
good point. A bare-bones version of the Jelly is good then.

File: src/Poisson.jl (hunk 1 above)
Comment (TzuYaoHuang) [reply src/Poisson.jl:89]: Maybe we could use ternary operator here? cuz `ifelse(cond, resTrue, resFalse)` will evaluate both `resTrue` and `resFalse` before entering the decision process, while `cond ? resTrue :  resFalse` will only evaluate the necessary one.

Comment (weymouth) [reply src/Poisson.jl:89]: This is what benchmarks are good for. According to the documentation `ifelse` avoids branch conditions. Which _sounds_ fast...

Comment (b-fg) [reply src/Poisson.jl:89]: I will setup the jelly benchmark and we see :)

Comment (weymouth) [reply src/Poisson.jl:89]: This change shouldn't reduce performance on the TGV either (as long as I haven't messed something up). It replaces all the σᵥ code with one `ifelse`.

**Comment by weymouth** [comment]
image
`5dc`=main branch and `5d3`=this PR. I think all those numbers are within the noise, but I will also try with the ternary @TzuYaoHuang suggested.

**Comment by weymouth** [comment]
Results essentially as expected. This is very slightly _worse_ when there is no body (since `?:` makes a branch we never use) and very slightly better when there is a body (since `?:` avoids the `mult` in body points). I think it's in the noise, but maybe the ternary wins? 
image

**Comment by b-fg** [comment]
Which one is `ifelse` and which on es the ternary?

**Comment by weymouth** [comment]
I also tried defining `zeroT=zero(eltype(p.r))` and using that in the `?:`, but the results were odd. It slightly sped up the TGV and slightly slowed down the Jelly, which is the opposite of what I would have expected. I think these benchmarks aren't really appropriate for micro-optimization of a single loop in the code. In this case, I just don't think it's going to matter enough, but in the future, micro-optimizations like this probably need to be decided with specialized code (ie, `@benchmark residual!(...)`).

**Comment by weymouth** [comment]
> Which one is `ifelse` and which on es the ternary?

The first test uses the PR as submitted. The second test uses the PR with this change:
```
- function residual!(p::Poisson) 
-     @inside p.r[I] = ifelse(p.iD[I]==0,0,p.z[I]-mult(I,p.L,p.D,p.x))
+ function residual!(p::Poisson, zeroT=zero(eltype(p.r)))
+    @inside p.r[I] = p.iD[I]==zeroT ? zeroT : p.z[I]-mult(I,p.L,p.D,p.x)
```

Maybe you could benchmark both versions. I'm not seeing a significant difference.

**Comment by b-fg** [comment]
Yes, I agree that micro-benchmarks for new functions are a must. This should be everyone's responsibility when including new code though. About `zeroT` I also wanted to test this by changing it across the entire code (whenever it makes sense). I will report back later on today.

**Comment by b-fg** [comment]
Oh also, in your tests, you changed two things at the same time right (ternary vs `ifelse`, and `zeroT`) ?

**Comment by weymouth** [comment]
No, I didn't add the `zeroT` thing into the second benchmark run above. I always try to do one thing at a time. ;-)

**Comment by b-fg** [comment]
Just runs all the tests:
 - 65a1: master
 - 5d3c: current PR (`ifelse`, no `zeroT`)
 - c556: use ternary (only)
 - 17bb: use `zeroT` (only)
 - 21fd: use both ternary and `zeroT`
 
Focusing on the bigger cases (`p=7`), the TGV shows that the differences are pretty small, so no real conclusions can be drawn. The jelly case shows that using `zeroT` yields a repeatable gain (it speeds the current PR up around 9% for both `ifelse` and ternary tests), and brings it closer to master. So I suggest including the `zeroT` change with the current version of the conditional (`ifelse`).

**Comment by b-fg** [comment]
Uh well, I repeated some tests just in case and the "gains" are no longer there. So you are right, is hard to catch these small changes with the big benchmarks.

**Comment by weymouth** [comment]
Let's just leave it as is then. We can do another PR with sweeping zeroT changes later is you like. As far as `≈` goes, you can't use it for values near zero (it uses a percent of the magnitude to do comparisons, so it fails when comparing near zero) and in this case it really isn't needed because I set `iD=0` exactly in a robust way when setting up the matrix.

**Pull request merged**